File: 알고리즘/볼트-너트맞추기.cpp
#include <bits/stdc++.h>
using namespace std;

// 입력 배열: Bolts, Nuts
// Bolts-Bolts/Nuts-Nuts 비교는 할 수 없음
vector<int> Bolts, Nuts;
// 볼트와 너트의 비교: Bolts[i] vs Nuts[j]
// 반환값:
// (1) -1 if Bolts[i] < Nuts[j]
// (2) 0 if Bolts[i] == Nuts[j]
// (3) +1 if Bolts[i] > Nuts[j]
int CmpBoltNut(int i, int j)
{
    if (Bolts[i] < Nuts[j])
        return -1;
    if (Bolts[i] > Nuts[j])
        return +1;
    return 0;
}

void swapArr(vector<int> &arr, int i, int j)
{
    int temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
}

void calc(vector<int> &bolts, vector<int> &nuts, int low, int high, vector<int> &boltNum, vector<int> &nutNum)
{
    if (low >= high)
        return;

    int j = low;
    int target = low;

    for (int i = low; i <= high; i++)
    {
        if (CmpBoltNut(low, i) == 0)
        {
            target = i;
        }
        if (CmpBoltNut(low, i) >= 0)
        {
            swapArr(Nuts, i, j);
            swapArr(nutNum, i, j);
            if (Bolts[low] == nuts[j])
            {
                target = j;
            }
            j++;
        }
    }
    j--;

    swapArr(Nuts, target, j);
    swapArr(nutNum, target, j);
    int k = low;

    for (int i = low; i <= high; i++)
    {
        if (CmpBoltNut(i, j) == 0)
        {
            target = i;
        }
        if (CmpBoltNut(i, j) <= 0)
        {
            swapArr(Bolts, i, k);
            swapArr(boltNum, i, k);
            if (Bolts[k] == nuts[j])
            {
                target = k;
            }
            k++;
        }
    }
    k--;
    swapArr(Bolts, target, k);
    swapArr(boltNum, target, k);
    calc(Bolts, Nuts, low, j - 1, boltNum, nutNum);
    calc(Bolts, Nuts, j + 1, high, boltNum, nutNum);
    return;
}

int main()
{
    int num_test_cases;
    cin >> num_test_cases;
    while (num_test_cases--)
    {
        int n; // 볼와 너트의 개수
        cin >> n;
        vector<int> boltNum(n + 1);
        vector<int> nutNum(n + 1);
        Bolts.assign(n + 1, 0);
        Nuts.assign(n + 1, 0);
        for (int i = 1; i <= n; ++i)
        {
            boltNum[i] = i;
            cin >> Bolts[i];
        }
        for (int j = 1; j <= n; ++j)
        {
            nutNum[j] = j;
            cin >> Nuts[j];
        }
        // 볼트와 너트 문제 해결 알고리즘
        // 볼트와 너트의 크기를 비교하기 위해 함수 CmpBoltNut 를 호출해야 함
        // 다음과 같이 볼트끼리 혹은 너트끼리 비교할 수 없음
        // if (Bolts[i] == Bolts[j]) 비교 불가능함
        // if (Nuts[i] == Nuts[j]) 비교 불가능함
        calc(Bolts, Nuts, 0, n, boltNum, nutNum);
        vector<int> answer(n + 1);
        for (int i = 0; i < boltNum.size(); i++)
        {
            answer[boltNum[i]] = nutNum[i];
        }
        // cout << "answer"<<endl;
        for (int i = 1; i < answer.size(); i++)
        {
            cout << answer[i]<< " ";
        }
        cout<<endl;
    }
    return 0;
}

File: 알고리즘/볼트-너트맞추기.py
import sys

input = sys.stdin.readline
sys.setrecursionlimit(1_000_000)


# 볼트와 너트의 비교: bolts[i] vs nuts[j]
# 반환값:
# (1) -1 if bolts[i] < nuts[j]
# (2) 0 if bolts[i] == nuts[j]
# (3) +1 if bolts[i] > nuts[j]


def cmp_bolt_nut(bolts, nuts, i, j):
    if bolts[i] < nuts[j]:
        return -1
    elif bolts[i] > nuts[j]:
        return +1
    else:
        return 0


def swap(arr, i, j):
    temp = arr[i]
    arr[i] = arr[j]
    arr[j] = temp
    return


def calc(bolts, nuts, low, high, bolt_num, nut_num):
    if low >= high:
        return
    j = low
    target = low

    # 피벗이 볼트[low]임
    # 너트 정렬
    # print(f"pivot : {bolts[low]}")
    for i in range(low, high + 1):
        if cmp_bolt_nut(bolts=bolts, nuts=nuts, i=low, j=i) == 0:  # pivot == nut[i]
            target = i
        if cmp_bolt_nut(bolts=bolts, nuts=nuts, i=low, j=i) in [1, 0]:  # pivot > nut[i]

            # swap(nuts, i, j)
            nuts[i], nuts[j] = nuts[j], nuts[i]
            # swap(nut_num, i, j)
            nut_num[i], nut_num[j] = nut_num[j], nut_num[i]
            if bolts[low] == nuts[j]:
                target = j
            j += 1

    j -= 1

    # swap(nuts, target, j)
    nuts[target], nuts[j] = nuts[j], nuts[target]
    # swap(nut_num, target, j)
    nut_num[target], nut_num[j] = nut_num[j], nut_num[target]

    # print(f"nuts: {nuts}")
    # print(f"nuts_num: {nut_num}")
    k = low
    # 피벗이 볼트의 low 였던 nuts[j]임
    # 볼트 정렬

    for i in range(low, high + 1):
        if cmp_bolt_nut(bolts=bolts, nuts=nuts, i=i, j=j) == 0:  # pivot == nut[i]
            target = i
        if cmp_bolt_nut(bolts=bolts, nuts=nuts, i=i, j=j) in [-1, 0]:  # pivot > nut[i]
            # swap(bolts, i, k)
            bolts[i], bolts[k] = bolts[k], bolts[i]
            # swap(bolt_num, i, k)
            bolt_num[i], bolt_num[k] = bolt_num[k], bolt_num[i]
            if bolts[k] == nuts[j]:
                target = k
            k += 1
    k -= 1
    # print(f"target:{target}, k:{k}")
    # swap(bolts, target, k)
    bolts[target], bolts[k] = bolts[k],bolts[target]
    # swap(bolt_num, target, k)
    bolt_num[target], bolt_num[k] = bolt_num[k],bolt_num[target]
    # print(f"bolts: {bolts}")
    # print(f"bolt_num: {bolt_num}")
    # print(low, j, high)
    calc(
        bolts=bolts, nuts=nuts, low=low, high=j - 1, bolt_num=bolt_num, nut_num=nut_num
    )
    calc(
        bolts=bolts, nuts=nuts, low=j + 1, high=high, bolt_num=bolt_num, nut_num=nut_num
    )

    return


t = int(input())

for _ in range(t):
    n = int(input())
    bolts = list(map(int, input().split()))
    bolt_num = [i for i in range(n)]
    nuts = list(map(int, input().split()))
    nut_num = [i for i in range(n)]

    qwe = [0 for _ in range(n)]
    calc(bolts, nuts, 0, n - 1, bolt_num, nut_num)

    answer = [0 for _ in range(n)]
    for i in range(n):
        answer[bolt_num[i]] = nut_num[i] + 1
    print(" ".join(map(str, answer)))
